lib/spack/spack/util/git.py

```diff
@@ -40,3 +41,25 @@ def get_modified_files(from_ref: str = "HEAD~1", to_ref: str = "HEAD") -> List[s
     stdout = git_exe("diff", "--name-only", from_ref, to_ref, output=str)
 
     return stdout.split()
+
+
+def get_commit_sha(path: str, ref: str) -> Optional[str]:
+    """Get a commit sha for an arbitrary ref"""
+    sha = None
+
+    # search for matching branch, then tag
+    ref_list = [f"refs/heads/{ref}", f"refs/tags/{ref}"]
+
+    if os.path.isdir(path):
+        # for the filesystem an unpacked mirror could be in a detached state from a depth 1 clone
+        # only reference there will be HEAD
+        ref_list.append("HEAD")
+
+    for try_ref in ref_list:
+        query = git(required=True)("ls-remote", path, try_ref, output=str, error=str)
+
+        if query:
+            sha = query.strip().split()[0]
+            break
+
+    return sha
```

Comment:
```suggestion
            return query.strip().split()[0]

    return None
```
And then we can get rid of the `sha` variable from the top of the method as well. Technically the `return None` could be left implicit, but I think it's better to keep it explicit for legibility.